api/models/Tag.js

```diff
@@ -0,0 +1,105 @@
+var tagsInText = (text = '') => {
+  // TODO alphanumeric and underscore
+  return (text.match(/#\w+/g) || []).map(str => str.substr(1, str.length))
+}
+
+var addToTaggable = (taggable, tagName, selected, trx) => {
+  var association, communities
+  var isPost = !taggable.post
+  if (isPost) {
+    // taggable is a post
+    association = 'communities'
+    communities = post => post.relations.communities.models
+  } else {
+    // taggable is a comment
+    association = 'post.communities'
+    communities = comment => comment.relations.post.relations.communities.models
+  }
+  return taggable.load(association)
+  .then(() => Tag.find(tagName))
+  .then(tag => {
+    if (tag) {
+      return tag
+    } else {
+      return new Tag({name: tagName}).save({}, {transacting: trx})
+    }
+  })
+  .tap(tag => {
+    var attachment = {tag_id: tag.id}
+    if (isPost) {
+      attachment.selected = selected
+    }
+    return taggable.tags().attach(attachment, {transacting: trx})
+  })
+  .then(tag => Promise.map(communities(taggable), com => addToCommunity(com, tag, taggable.user_id, trx)))
+}
+
+var removeFromTaggable = (taggable, tag, trx) => {
+  return taggable.tags().detach(tag.id, {transacting: trx})
+}
+
+var addToCommunity = (community, tag, user_id, trx) => {
+  return CommunityTag.where({community_id: community.id, tag_id: tag.id}).fetch()
```

Comment:
this needs to take place inside a transaction otherwise it's subject to a race condition